docs/howto/enable-dag-serialization.rst

```diff
@@ -0,0 +1,109 @@
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+
+
+
+Enable DAG Serialization
+========================
+
+Add the following settings in ``airflow.cfg``:
+
+.. code-block:: ini
+
+    [core]
+    store_serialized_dags = True
+    min_serialized_dag_update_interval = 30
+
+*   ``store_serialized_dags``: This flag decides whether to serialises DAGs and persist them in DB.
+    If set to True, Webserver reads from DB instead of parsing DAG files
+*   ``min_serialized_dag_update_interval``: This flag sets the minimum interval (in seconds) after which
+    the serialized DAG in DB should be updated. This helps in reducing database write rate.
+
+If you are updating Airflow from <1.10.6, please do not forget to run ``airflow db upgrade``.
+
+
+How it works
+------------
+
+In order to make Airflow Webserver stateless (almost!), Airflow >=1.10.6 supports
+DAG Serialization and DB Persistence.
+
+.. image:: ../img/dag_serialization.png
+
+As shown in the image above in Vanilla Airflow, the Webserver and the Scheduler both
+needs access to the DAG files. Both the scheduler and webserver parses the DAG files.
+
+With **DAG Serialization** we aim to decouple the webserver from DAG parsing
+which would make the Webserver very light-weight.
+
+As shown in the image above, when using the dag_serilization feature,
+the Scheduler parses the DAG files, serializes them in JSON format and saves them in the Metadata DB.
+
+The Webserver now instead of having to parse the DAG file again, reads the
+serialized DAGs in JSON, de-serializes them and create the DagBag and uses it
+to show in the UI.
+
+One of the key features that is implemented as the part of DAG Serialization is that
+instead of loading an entire DagBag when the WebServer starts we only load each DAG on demand from the
+Serialized Dag table. This helps reduce Webserver startup time and memory. The reduction is notable
+when you have large number of DAGs.
+
+Below is the screenshot of the ``serialized_dag`` table in Metadata DB:
+
+.. image:: ../img/serialized_dag_table.png
+
+Limitations
+-----------
+The Webserver will still need access to DAG files in the following cases,
+which is why we said "almost" stateless.
+
+*   **Rendered Template** tab will still have to parse Python file as it needs all the details like
+    the execution date and even the data passed by the upstream task using Xcom.
+*   **Code View** will read the DAG File & show it using Pygments.
+    However, it does not need to Parse the Python file so it is still a small operation.
+*   :doc:`Extra Operator Links <define_extra_link>` would not work out of
+    the box. They need to be defined in Airflow Plugin.
+
+    **Existing Airflow Operators**:
+    To make extra operator links work with existing operators like BigQuery, copy all
+    the classes that are defined in ``operator_extra_links`` property.
```

Comment:
Does it mean `operator_extra_links` can only be registered through plugin in the future?

Reply:
If dag serialization becomes the default yes. Well, either bundled in Airflow, or registered via a plugin.

@kaxil We should check that we serialize the "bundled" GCP links correctly (they should work as the webserver will have the code for this)

Reply:
Can you clarify a bit about the "bundled" GCP links? Are you talking about the operator extra links registered as a `@property` in the operator like:
https://github.com/apache/airflow/blob/master/airflow/gcp/operators/bigquery.py#L460-L470
?

Reply:
Yes, unfortunately in the current implementation you need to add operator links via plugin for in-built operators too.

Let me try and see if I can remove the need to register links for in-built operators links

Reply:
Looks like the currently defined OperatorLinks for inbuilt operators aredependent on Instance of operator object.

Example the following is Qubole Operator:

```
class QDSLink(BaseOperatorLink):
    """Link to QDS"""
    name = 'Go to QDS'

    def get_link(self, operator, dttm):
        return operator.get_hook().get_extra_links(operator, dttm)
```
which depends on `conn = BaseHook.get_connection(operator.kwargs['qubole_conn_id'])` line. We don't store `qubole_conn_id` as the task property in our serialized DAG.

and the BigqueryOperator is:

```
@property
    def operator_extra_links(self):
        """
        Return operator extra lxinks
        """
        if isinstance(self.sql, str):
            return (
                BigQueryConsoleLink(),
            )
        return (
            BigQueryConsoleIndexableLink(i) for i, _ in enumerate(self.sql)
        )
```
this one is dependent on `self.sql` property !! which the Serialized Operator won't have access to as we don't store that info in serialized DAG.

Reply:
Hmmmm. Tricky. And even if we did store the attributes (which is doable by setting `_serialized_fields` on the operator sub-classes) these links might not work for Quoble: `operator` will be SerializedBaseOperator, not a `QuobleOperator`.

(BQ would work well enough as the SQL would be of the right length, even if it was templated so the value stored in s10n is not the final value)

Reply:
Exactly. Hence I feel to let the user create their own OperatorLink via plugin !

And storing properties like `self.sql` will increase the serialized blob size as it is a templated field it can take in a huge SQL Query.

Reply:
Should we create a pre-existing plugin and register the existing links there so that users can re-use them easily?